homeassistant/components/sonos/sensor.py

```diff
@@ -224,3 +244,89 @@ async def _async_update_state(self) -> None:
+    """Representation of a Sonos Next Alarm entity."""
+
+    _attr_device_class = SensorDeviceClass.TIMESTAMP
+    _attr_has_entity_name = True
+    _attr_translation_key = "next_alarm"
+
+    def __init__(self, speaker: SonosSpeaker, config_entry: SonosConfigEntry) -> None:
+        """Initialize the next alarm sensor."""
+        super().__init__(speaker, config_entry)
+        self._attr_unique_id = f"{self.soco.uid}-next-alarm"
+        self.household_id = speaker.household_id
+
+    async def async_added_to_hass(self) -> None:
+        """Handle sensor setup when added to hass."""
+        await super().async_added_to_hass()
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{SONOS_ALARMS_UPDATED}-{self.household_id}",
+                self.async_write_ha_state,
+            )
+        )
+
+        async def async_write_state_daily(now: datetime.datetime) -> None:
+            """Update state attributes each calendar day."""
+            _LOGGER.debug("Updating state attributes for %s", self.name)
+            self.async_write_ha_state()
+
+        self.async_on_remove(
+            async_track_time_change(
+                self.hass, async_write_state_daily, hour=0, minute=0, second=0
+            )
+        )
+
+    async def _async_fallback_poll(self) -> None:
+        """Call the central alarm polling method."""
+        alarms: SonosAlarms = self.config_entry.runtime_data.alarms[self.household_id]
+        assert alarms.async_poll
+        await alarms.async_poll()
+
+    @property
+    def icon(self) -> str | None:
+        """Icon of the entity."""
+        if self.native_value is None:
+            return "mdi:alarm-off"
+        if self._following_alarm is None:
+            return "mdi:alarm"
+        return "mdi:alarm-multiple"
+
+    @property
+    def native_value(self) -> datetime.datetime | None:
+        """Return the state of the sensor."""
+        return self.speaker.alarms.get_next_alarm_datetime(zone_uid=self.soco.uid)
+
+    @property
+    def _is_today(self) -> bool:
+        """Return whether this next alarm is scheduled for today."""
+        if self.native_value is None:
+            return False
+        return self.native_value.date() == datetime.datetime.today().date()
+
+    @property
+    def _following_alarm(self) -> datetime.datetime | None:
+        """Return the following alarm after the next current one."""
+        if self.native_value is None:
+            return None
+        return self.speaker.alarms.get_next_alarm_datetime(
+            from_datetime=self.native_value + datetime.timedelta(minutes=1),
+            zone_uid=self.soco.uid,
+        )
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return attributes of Sonos next alarm sensor."""
+        return {
+            ATTR_SCHEDULED_TODAY: self._is_today,
+            ATTR_FOLLOWING_ALARM: self._following_alarm,
+        }
```

Comment:
Should the entity_id of the next alarm be in here as an attribute?

Reply:
The function in SoCo only returns the next time, not any other attributes. I can later open a pull in SoCo to improve this functionality.

Reply:
got it. sounds fine.